app/src/main/java/com/nutomic/syncthingandroid/model/Completion.java

```diff
@@ -0,0 +1,138 @@
+                break;
+            }
+        }
+    }
+
+    /**
+     * Updates device and folder information in the cache model
+     * after a config update.
+     */
+    public void updateFromConfig(List<Device> newDevices, List<Folder> newFolders) {
+        // Handle devices that were removed from the config.
+        List<String> removedDevices = new ArrayList<>();;
+        Boolean deviceFound;
+        for (String deviceId : deviceFolderMap.keySet()) {
+            deviceFound = false;
+            for (Device device : newDevices) {
+                if (device.deviceID.equals(deviceId)) {
+                    deviceFound = true;
+                    break;
+                }
+            }
+            if (!deviceFound) {
+                removedDevices.add(deviceId);
+            }
+        }
+        for (String deviceId : removedDevices) {
+            Log.v(TAG, "updateFromConfig: Remove device '" + deviceId + "' from cache model");
+            deviceFolderMap.remove(deviceId);
+        }
+
+        // Handle devices that were added to the config.
+        for (Device device : newDevices) {
+            if (!deviceFolderMap.containsKey(device.deviceID)) {
+                Log.v(TAG, "updateFromConfig: Add device '" + device.deviceID + "' to cache model");
+                deviceFolderMap.put(device.deviceID, new DefaultHashMap<String, CompletionInfo>());
+            }
+        }
+
+        // Handle folders that were removed from the config.
+        List<String> removedFolders = new ArrayList<>();;
+        Boolean folderFound;
+        for (String deviceId : deviceFolderMap.keySet()) {
+            for (String folderId : deviceFolderMap.get(deviceId).keySet()) {
+                folderFound = false;
+                for (Folder folder : newFolders) {
+                    if (folder.id.equals(folderId)) {
+                        folderFound = true;
+                        break;
+                    }
+                }
+                if (!folderFound) {
+                    removedFolders.add(folderId);
+                }
+            }
+        }
+        for (String folderId : removedFolders) {
+            Log.v(TAG, "updateFromConfig: Remove folder '" + folderId + "' from cache model");
+            removeFolder(folderId);
+        }
+
+        // Handle folders that were added to the config.
+        for (Folder folder : newFolders) {
+            for (Device device : newDevices) {
+                if (folder.getDevice(device.deviceID) != null) {
+                    // folder is shared with device.
+                    if (!deviceFolderMap.get(device.deviceID).containsKey(folder.id)) {
+                        Log.v(TAG, "updateFromConfig: Add folder '" + folder.id +
+                                    "' shared with device '" + device.deviceID + "' to cache model.");
+                        deviceFolderMap.get(device.deviceID).put(folder.id, new CompletionInfo());

```

Comment:
Doing the get twice just seems wrong. Assign it to some variable somewhere and then do a put.

Reply:
ok